From 39f09114b425a248516f27296a57b91978d47b81 Mon Sep 17 00:00:00 2001
From: Jonathan Kew <jkew@mozilla.com>
Date: Sun, 9 Oct 2016 09:18:37 -0700
Subject: [PATCH 3/6] Trac 20206: Avoid the need to download the font Osaka

 * Bug 1283573 - Don't attempt to look up a single-face-family name if
   the corresponding main family isn't available. r=mstange
---
 gfx/thebes/gfxMacPlatformFontList.mm | 17 +++++++++++++----
 modules/libpref/init/all.js          |  6 +++---
 2 files changed, 16 insertions(+), 7 deletions(-)

diff --git a/gfx/thebes/gfxMacPlatformFontList.mm b/gfx/thebes/gfxMacPlatformFontList.mm
index eb7a788..5c63e69 100644
--- a/gfx/thebes/gfxMacPlatformFontList.mm
+++ b/gfx/thebes/gfxMacPlatformFontList.mm
@@ -735,15 +735,24 @@ gfxMacPlatformFontList::InitSingleFaceList()
     for (uint32_t i = 0; i < numFonts; i++) {
         LOG_FONTLIST(("(fontlist-singleface) face name: %s\n",
                       NS_ConvertUTF16toUTF8(singleFaceFonts[i]).get()));
-        if (!gfxFontUtils::IsFontFamilyNameAllowed(singleFaceFonts[i])) {
+        nsAutoString familyName(singleFaceFonts[i]);
+        auto colon = familyName.FindChar(':');
+        if (colon != kNotFound) {
+            nsAutoString key(Substring(familyName, colon + 1));
+            ToLowerCase(key);
+            if (!mFontFamilies.GetWeak(key)) {
+                continue;
+            }
+            familyName.Truncate(colon);
+        }
+        if (!gfxFontUtils::IsFontFamilyNameAllowed(familyName)) {
             continue;
         }
-        gfxFontEntry *fontEntry = LookupLocalFont(singleFaceFonts[i],
+        gfxFontEntry *fontEntry = LookupLocalFont(familyName,
                                                   400, 0,
                                                   NS_FONT_STYLE_NORMAL);
         if (fontEntry) {
-            nsAutoString familyName, key;
-            familyName = singleFaceFonts[i];
+            nsAutoString key;
             GenerateFontListKey(familyName, key);
             LOG_FONTLIST(("(fontlist-singleface) family name: %s, key: %s\n",
                           NS_ConvertUTF16toUTF8(familyName).get(),
diff --git a/modules/libpref/init/all.js b/modules/libpref/init/all.js
index e855cdb..2891c80 100644
--- a/modules/libpref/init/all.js
+++ b/modules/libpref/init/all.js
@@ -3577,9 +3577,9 @@ pref("font.name.monospace.x-math", "Courier");
 pref("font.name.cursive.x-math", "Apple Chancery");
 pref("font.name.fantasy.x-math", "Papyrus");
 
-// individual font faces to be treated as independent families
-// names are Postscript names of each face
-pref("font.single-face-list", "Osaka-Mono");
+// Individual font faces to be treated as independent families,
+// listed as <Postscript name of face:Owning family name>
+pref("font.single-face-list", "Osaka-Mono:Osaka");
 
 // optimization hint for fonts with localized names to be read in at startup, otherwise read in at lookup miss
 // names are canonical family names (typically English names)
-- 
2.10.2

